### ISSUE TITLE
`where: { shop: undefined }`: The entire table data was accidentally deleted

### ISSUE THREAD

**Issue by bithaolee** [prisma/prisma#20169]
### Bug description

```nodejs
let shop = undefined; // shop is variable, and in some cases, it may be undefined

await prisma.theme.deleteMany({
  where: {
    shop,
  }
})
```
I left out the shop parameter check, causing the shop to be undefined in some cases, triggering the deletion of the entire table. This operation of Prisma is very risky, I hope to improve it.

**Comment by millsp** [comment]
Hey, thanks for the report, we are aware of this issue and I fully agree with you that it can be dangerous. We should improve it.

**Comment by bithaolee** [comment]
Thank you  @millsp,  maybe updateMany operation has a similar issue that I haven't tested, and if so, I'd like to see it solved together

**Comment by janpio** [comment]
It is not an issue, but how Prisma is currently designed to work: https://www.prisma.io/docs/concepts/components/prisma-client/null-and-undefined We know this can be dangerous, but it has been like this for quite some while and switching would also have other deficits.

So as soon as we find the proper issue that requests us to change this, we will close this issue here as a duplicate.

**Comment by why-el** [comment]
This is extremely dangerous, I do not think it's being treated as such. You are offering an abstraction that does the opposite of what the SQL standard does (and the majority of your users understand SQL before they understand Prisma). When filtering by emptiness (`where id in []`, or undefined), no database will return records, but Prisma does. This is even worse in the case of deletes. 

I fear this is one famous case of mass-deletion away from a PR disaster for Prisma. I am a happy user of the library, mind you, but I'd recommend more urgency to this issue. I spent 2 minutes googling, and I already found a case of someone who had to suffer from this issue: https://www.farmin.dev/posts/35ad668d-29ae-4d86-8b4a-126b6d561aaa.

**Comment by lnfel** [comment]
Dude I just shot myself on the foot right now. As a Prisma advocate, my spirit is in shambles right this moment.

**Comment by lnfel** [comment]
I don't care if breaking change were to be introduced, this evil shall be corrected.

**Comment by lnfel** [comment]
Apparently if the data we are querying is expected to be `undefined` for some reason we can use nullish coalescing operator  just like:
```js
// this will return null back to us
const user = await prisma.user.findFirst({
  where: {
    // let's use an empty string instead and prisma will use it in the SQL query!
    id: somethingNull?.id ?? '',
  }
})
```
OP's issue can be avoided by doing the same thing `shop ?? ''`.
My spirit is now in shambles.

**Comment by jakezegil** [comment]
We unfortunately ran into this issue (luckily had snapshots we could restore), and I had to write some middleware because this was super dangerous. I hope this can help somebody!

```ts
import { isEmpty } from "lodash";
import { Prisma } from "prisma/generated";

/**
 * Middleware featured below
 */

// This naming convention specifically matches the
// safe word to the exported constant to support the
// middleware error handling
export const safeWordDelete = "FORCE_DELETE_ALL";
export const FORCE_DELETE_ALL = { id: safeWordDelete };

export const safeWordFind = "FIND_ALL";
export const FIND_ALL = { id: safeWordFind };

export const safeWordUpdate = "UPDATE_ALL";
export const UPDATE_ALL = { id: safeWordUpdate };

/**
 * If you have found yourself here after a hard lesson:
 *
 * These middleware specifically do NOT guard in the case of a nested find, update, delete.
 * These middleware also don't protect in the case of an undefined variable being passed in alongside a specific condition.
 *
 * Ex: {
 *    id, // undefined
 *    otherAttr: null
 * }
 *
 * The above where clause will delete/modify/find all regardless of undefined status of variable appId.
 *
 * The !! best !! guard here is proper type adherence - do not bang, do not coerce.
 */
const _guardEnMasse = ({
  params,
  actions,
  safeWord,
}: {
  params: Prisma.MiddlewareParams;
  actions: Prisma.MiddlewareParams["action"][];
  safeWord: string;
}) => {
  // Check if empty, if not and safeWord is provided, then set where = {} and proceed to execute
  if (actions.includes(params.action)) {
    // Don't allow delete all queries
    if (isEmpty(params.args.where)) {
      throw new Error(
        `It looks like you just tried to perform a ${params.action} on all of the ${params.model}s. If this was intentional, pass 'where: ${safeWord}'`,
      );
    }
  }

  if (params.args?.where?.id === safeWord) {
    params.args.where = {};
  }
};

export const guardEnMasse: PrismaMiddleware = async (params, next) => {
  /* DELETION PROTECTION MIDDLEWARE */
  _guardEnMasse({
    actions: ["delete", "deleteMany"],
    params,
    safeWord: safeWordDelete,
  });

  /* MASS FIND PROTECTION MIDDLEWARE */
  _guardEnMasse({
    actions: ["findMany"],
    params,
    safeWord: safeWordFind,
  });

  /* MASS UPDATE PROTECTION MIDDLEWARE */
  _guardEnMasse({
    actions: ["updateMany"],
    params,
    safeWord: safeWordUpdate,
  });

  return next(params);
};
```

and then 
```ts
const dbClient = new PrismaClient();

coreDBClient.$use(guardEnMasse);
```

Best guard here is strict type safety and attribute validation ahead of db writes. Also, always always always backup your production database!

**Comment by scumbkt19** [comment]
Thank you @jakezegil for sharing how to add runtime safety to these queries, but I wondered if more could be done to prevent this sort of issue at compile time since all our infrastructure would be happy shipping `dbClient.sensitiveData.deleteMany()` to prod.

Heavily influenced by @kz-d's https://github.com/kz-d/prisma-where-required generator, I wrote the following generator which 1) makes the argument to all the bulk operations required and then 2) makes the `where` clause in the parameter required.

Combining the buildtime and runtime checks provides a degree of safety I feel more comfortable with.

Setup is the same as in `prisma-where-required` but the actual source code was modified to

generator.ts
```typescript
#!/usr/bin/env node

import { generatorHandler } from "@prisma/generator-helper";
import { logger } from "@prisma/internals";

import { convert } from "./convertor";

generatorHandler({
  onManifest: () => ({
    prettyName: "where-required",
    requiresGenerators: ["prisma-client-js"],
    defaultOutput: ".",
  }),
  onGenerate: async (options) => {
    const nodeModulePath = options.generator.config.nodeModulePath;
    if (!nodeModulePath) {
      logger.error("No nodeModulePath config.");
      return;
    }

    const debug = options.generator.config.debug === "true";

    const models = options.dmmf.datamodel.models;

    convert({
      path: `${nodeModulePath}/.prisma/client/index.d.ts`,
      models: models.map((model) => model.name),
      debug,
    });
  },
});
```

convertor.ts
```typescript
import { Project, SourceFile, ts } from "ts-morph";
import isPropertySignature = ts.isPropertySignature;

function removeQuestionTokenFromWhereFieldProperty(
  sourceFile: SourceFile,
  models: string[]
) {
  const prismaNamespace = sourceFile.getModuleOrThrow("Prisma");
  let isInTypeAlias = false;
  prismaNamespace.transform((tr) => {
    const currentNode = tr.currentNode;
    if (
      ts.isTypeAliasDeclaration(currentNode) &&
      models.some(
        (targetModel) =>
          currentNode.name.getText().startsWith(targetModel) &&
          !currentNode.name.getText().startsWith(`${targetModel}$`) &&
          (currentNode.name.getText().endsWith("FindManyArgs") ||
            currentNode.name.getText().endsWith("DeleteManyArgs") ||
            currentNode.name.getText().endsWith("DeleteArgs") ||
            currentNode.name.getText().endsWith("UpdateManyArgs"))
      )
    ) {
      isInTypeAlias = true;
      const newNode = tr.visitChildren();
      isInTypeAlias = false;
      return newNode;
    }
    if (
      isInTypeAlias &&
      isPropertySignature(currentNode) &&
      currentNode.name.getText() === "where"
    ) {
      isInTypeAlias = false;
      return tr.factory.updatePropertySignature(
        currentNode,
        currentNode.modifiers,
        currentNode.name,
        undefined,
        currentNode.type
      );
    }
    return tr.visitChildren();
  });
}

function removeQuestionTokenFromActionFunctionProperty(
  sourceFile: SourceFile,
  models: string[]
) {
  const prismaNamespace = sourceFile.getModuleOrThrow("Prisma");

  let isInDelegate = false;
  let isInFindMany = false;
  prismaNamespace.transform((tr) => {
    const currentNode = tr.currentNode;
    if (
      ts.isInterfaceDeclaration(currentNode) &&
      models.some(
        (targetModel) => currentNode.name.getText() === `${targetModel}Delegate`
      )
    ) {
      isInDelegate = true;
      const newNode = tr.visitChildren();
      isInDelegate = false;
      return newNode;
    }
    if (
      isInDelegate &&
      ts.isMethodSignature(currentNode) &&
      currentNode.name.getText() === "findMany"
    ) {
      isInFindMany = true;
      const newNode = tr.visitChildren();
      isInFindMany = false;
      return newNode;
    }
    if (isInFindMany && ts.isParameter(currentNode)) {
      isInDelegate = false;
      isInFindMany = false;
      return tr.factory.updateParameterDeclaration(
        currentNode,
        currentNode.modifiers,
        currentNode.dotDotDotToken,
        currentNode.name,
        undefined,
        currentNode.type,
        currentNode.initializer
      );
    }
    return tr.visitChildren();
  });
}

export function convert(arguments_: {
  path: string;
  models: string[];
  debug: boolean;
}) {
  const project = new Project();
  const sourceFile = project.addSourceFileAtPath(arguments_.path);

  removeQuestionTokenFromWhereFieldProperty(sourceFile, arguments_.models);
  removeQuestionTokenFromActionFunctionProperty(sourceFile, arguments_.models);

  sourceFile.saveSync();
}

```
(I'll confess I found the ts-morph transform visitor difficult to reason about, there might be a more efficient way to write that loop)

this added <5 seconds to the `prisma generate` time for my 100+ model file

so then
```typescript

// Compile error (missing an arg)
dbClient.sensitiveData.deleteMany()

// Compile error (missing a where)
dbClient.sensitiveData.findMany({select: {...}})

// Runtime error (as above)
dbClient.sensitiveData.findMany({ where: {id: undefined}})

// Success via runtime override with maximal clarity on what's happening
dbClient.sensitiveData.findMany({ where: FIND_ALL })
```

**Comment by sorenbs** [comment]
I am seeking input on a proposal for changing this behavior.

Changing this will be a breaking change, which is fine. We just want to make sure we land on the right solution, and have a migration path in place for people who depend on the existing behavior.

# The proposal

Currently, `null` and `undefined` are treated differently by Prisma. This is convenient because it allows you to programmatically build up the where condition and skip certain filters by simply setting them to undefined. As this issue points out, it also introduces the chance for a simple programming error to lead to a query that is too broad - which is especially problematic for `deleteMany` queries.

The proposal is to change the behavior of `undefined` so it is exactly the same as `null` - it will filter for the null value in the database.

For situations where the where condition is being constructed programatically, you will now have to delete the property instead of setting it to undefined.

Instead of 

```
{
  age: 26,
  city: undefined
}
```

The where condition will have to be

```
{
  age: 26
}
```

To ease the transition for large projects that depend on the current behavior, we can introduce a client extension that restores the original behavior:

```
const prisma = new PrismaClient().extends(restoreOriginalUndefinedBehavior())
```

Please provide input to this proposal 🙏

**Comment by bithaolee** [comment]
> I am seeking input on a proposal for changing this behavior.
> 
> Changing this will be a breaking change, which is fine. We just want to make sure we land on the right solution, and have a migration path in place for people who depend on the existing behavior.
> 
> # The proposal
> Currently, `null` and `undefined` are treated differently by Prisma. This is convenient because it allows you to programmatically build up the where condition and skip certain filters by simply setting them to undefined. As this issue points out, it also introduces the chance for a simple programming error to lead to a query that is too broad - which is especially problematic for `deleteMany` queries.
> 
> The proposal is to change the behavior of `undefined` so it is exactly the same as `null` - it will filter for the null value in the database.
> 
> For situations where the where condition is being constructed programatically, you will now have to delete the property instead of setting it to undefined.
> 
> Instead of
> 
> ```
> {
>   age: 26,
>   city: undefined
> }
> ```
> 
> The where condition will have to be
> 
> ```
> {
>   age: 26
> }
> ```
> 
> To ease the transition for large projects that depend on the current behavior, we can introduce a client extension that restores the original behavior:
> 
> ```
> const prisma = new PrismaClient().extends(restoreOriginalUndefinedBehavior())
> ```
> 
> Please provide input to this proposal 🙏

Thanks for your efforts, I think it's a good idea, looking forward to it!

**Comment by Scalahansolo** [comment]
> The proposal is to change the behavior of undefined so it is exactly the same as null - it will filter for the null value in the database.

`undefined` and `null` aren't the same though. While moving in this direction is theoretically safer, it's not as safe as it should be. Honestly, I believe the correct answer here is that Prisma should just throw an error if an undefined value is passed.

**Comment by SevInf** [comment]
Hello everyone, in the past weeks we've came up with a different proposal for handling that issue, addressing some of the feedback.

TL;DR: we'll introduce a preview feature flag for making explicitly `undefined` values invalid in Prisma 5, make it a default in Prisma 6, as well as introduce a safer shortcut for explicitly skipping part of the query.

Full proposal text below, we'd like to hear you comments on it:

# Problem

Currently, we treat `undefined` values as absent. Meaning, that any field set to `undefined` explicitly is treated as if the query was not supplied at all.

This makes some usecases, listed in the doc pretty convenient, however, it also causes problems: JS language itself is very inconsistent on when it treats `null` vs `undefined` as the same values or as two different ones and our behaviour is not the default most people expect.

More importantly, for write queries, the feature can accidentally cause a data loss: https://github.com/prisma/prisma/issues/20169

For read queries, similar situation might cause a data leak.

Overall, there is strong opinion that that feature was a bad decision on our part for a long time. This proposal suggests we change that behaviour.

## Proposal

### In Prisma 5

Introduce new preview feature `strictUndefinedChecks`. 

When preview feature is on, any value that is explicitly set to `undefined` will cause a runtime error.  Omitting a field must still be allowed - only explicit values are disallowed.

We adjust our doc to recommend setting `exactOptionalPropertyTypes` setting ([[docs](https://www.typescriptlang.org/tsconfig/#exactOptionalPropertyTypes)](https://www.typescriptlang.org/tsconfig/#exactOptionalPropertyTypes)) to get equivalent type error.

To simplify the migration, we also introduce `Prisma.skip` symbol which would act as “ignore this part of the query as if the field were not set” instead of `undefined`.

So these two calls are equivalent:

```ts
prisma.user.create({
   name: 'Steve the rat',
})

prisma.user.create({
   name: 'Steve the rat',
   profilePicture: Prisma.skip
})
```

In that case, migration from `undefined` to `Prisma.skip` is relatively simple:

```ts
let optionalProfilePicture: ProfilePicture | undefined

// before
prisma.user.create({
   name: 'Steve the rat',
   profilePicture: optionalProfilePicture
})

// after
prisma.user.create({
   name: 'Steve the rat',
   profilePicture: optionalProfilePicture ?? Prisma.skip
})
```

It is a safer option, because users are forced to be explicit about their intent.

### In Prisma 6

We remove `strictUndefinedChecks` preview feature and make this behavior a default. Depending on feedback, we might introduce a generator option to revert back to old, unsafe, behaviour.

**Comment by Scalahansolo** [comment]
This is a great solution along with the bonus of great migration path going forward. Glad to see this finally getting addressed!

**Comment by SevInf** [comment]
Hello everyone, thanks for the feedback on proposal.
Feature from above spec is implemented in #25224 and is available for testing as `5.20.0-integration-feat-strict-undefined-checks.1` snapshot version. We don't recommend using dev builds in production, but you could use it for test driving the feature before it will be published.

**Comment by drewcorlin1** [comment]
The release notes say that this will be the default in the next major version. While not a blocker, that will require many changes with how my team currently uses prisma. 

It's quite common for us to use prisma to implement REST GET endpoints to list entities that can have many optional query params. For that sort of use case it'll require us to make many changes to fallback to `Prisma.skip`. 

Is it possible to make this behavior configurable in the future?
